source/agora/consensus/PreImage.d

```diff
@@ -274,26 +274,17 @@ public struct PreImageCycle
 {
     /// Make sure we get initialized by disabling the default ctor
     @disable public this();
+
     /// Ditto
-    public this (typeof(PreImageCycle.tupleof) args)
+    public this (uint seed_interval)
     {
-        this.tupleof = args;
+        this.seeds = PreImageCache(PreImageCycle.SeedCount, seed_interval);
+        this.preimages = PreImageCache(seed_interval, 1);
     }
 
-    /// The number of cycles for a bulk of pre-images
-    public static immutable uint NumberOfCycles = 100;
-
-    /***************************************************************************
-
-        The number of the current cycle
-
-        This is the data used as a nonce in generating the cycle seed.
-        Named `nonce` to avoid any ambiguity. It is incremented once every
-        `EnrollPerCycle` period, currently 700 days.
-
-    ***************************************************************************/
-
-    public uint nonce;
+    /// The number of seeds for pre-images
+    /// If the cycle is 1008, about 5 million of pre-images will be generated.
+    public static immutable uint SeedCount = 5000;
```

Comment:
Am I missing something here ? Having a value which is dependent on the cycle will still create problems, won't it ?

Reply:
What is the `Seed` in this context?

Reply:
> What is the `Seed` in this context?

We can't store all the pre-images like 5 million pre-images, so we just maintain 5000 seeds. When a new set of pre-images, we can a set of pre-images from a proper seed as follows.
`
[preimage 2016] [preimage 2015] ... [preimage 2] [seed 2] [preimage 1008] [preimage 1007] ... [[preimage 2] [seed1]
`
If you have a more proper name for the variable, please let me know. :)

Reply:
I think the cache is supposed to define which `preImages` are stored in the array and not the `seed`. I think if we are going to use a continuous sequence of `preImages` then a target count is easier to understand and follow.

Reply:
Thank you for the detailed explanation :)
That is similar to `PreImageCycle` vs `PreImageCache`.

Reply:
> I think the cache is supposed to define which `preImages` are stored in the array and not the `seed`. I think if we are going to use a continuous sequence of `preImages` then a target count is easier to understand and follow.

I just use the `SeedCount` as a multiplier of a validator cycle because setting a total number of pre-images causes hashing 5 million times even in network tests or unit tests, which takes more than 1 second. That's why I used the seed count, not the total count.

Reply:
I thought if we needed to have it configurable for testing then we would make it configurable as `ConsensusParam` that we override in our unit tests. If you really want to stick with a multiplier of the `validatorCycle` then the original name `NumberOfCycles ` was more accurate.

Reply:
> Am I missing something here ? Having a value which is dependent on the cycle will still create problems, won't it ?

Yes, it will. I think I should change the code. My intention was just not to hash so many times in unit tests or network tests but it will leave the problem also. I will use 5 million preimages for the Coinnet and 100K preimages for the Testnet. How about this? @Geod24 

@hewison-chris I understand what you mean now. I was misunderstanding some point.